File: 프로그래머스/5/49190. 방의 개수/방의 개수.java
import java.util.*;

class Solution {
    
    static class Node {
        
        int x, y;
        String id;
        Set<String> connectedNode;

        public Node(int x, int y) {
            this.x = x;
            this.y = y;
            this.id = x + "," + y;
            this.connectedNode = new HashSet<>();
        }
    }
    
    public int solution(int[] arrows) {
        
        int[] dx = {-1, -1, 0, 1, 1, 1, 0, -1};
        int[] dy = {0, 1, 1, 1, 0, -1, -1, -1};
        
        Map<String, Node> map = new HashMap<>();
        
        Node prevNode = new Node(0, 0);
        map.put(prevNode.id, prevNode);
        
        int count = 0;

        for (int d : arrows) {
            for (int i = 0; i < 2; i++) {

                int nx = prevNode.x + dx[d];
                int ny = prevNode.y + dy[d];
                String newId = nx + "," + ny;

                //처음 이동한 좌표
                if (!map.containsKey(newId)) {
                    map.put(newId, new Node(nx, ny));
                }
                //이미 한번 이동한 좌표
                else if (!prevNode.connectedNode.contains(newId)) {
                    count++;
                }

                Node nextNode = map.get(newId);

                //두 좌표 연결
                prevNode.connectedNode.add(nextNode.id);
                nextNode.connectedNode.add(prevNode.id);

                prevNode = nextNode;
            }
        }                

        return count;
    }
}

File: 프로그래머스/5/49190. 방의 개수/방의 개수.py
def solution(arrows):
    
    dx = (-1, -1, 0, 1, 1, 1, 0, -1)
    dy = (0, 1, 1, 1, 0, -1, -1, -1)
    
    class Node:
        def __init__(self, x, y):
            self.x = x
            self.y = y
            self.id = f"{x},{y}"
            self.connected_node = set()
    
    count = 0
    map = {}
    v = Node(0, 0)
    map[v.id] = v
    
    for d in arrows:
        for _ in range(2):
            
            x = v.x + dx[d]
            y = v.y + dy[d]
            id = f"{x},{y}"
            
            if id not in map:
                map[id] = Node(x, y)
            elif id not in v.connected_node:
                count += 1
            
            u = map[id]
            
            v.connected_node.add(u.id)
            u.connected_node.add(v.id)
            
            v = u
    
    return count